Load your Deliverable 3 output

In [5]:
import pandas as pd

christchurch_all = pd.read_csv(
    "output/Christchurch_listings_Oct2025_Jun2026.csv"
)

christchurch_all.shape

(28795, 19)

Handle missing values ( In deliverable I found these are all missing values price, last_review, and reviews_per_month)

In [6]:
before_price = len(christchurch_all)

christchurch_all = christchurch_all.dropna(
    subset=["price"]
).copy()

after_price = len(christchurch_all)

print("Rows removed:", before_price - after_price)
print("Rows remaining:", after_price)

Rows removed: 10667
Rows remaining: 18128


In [7]:
christchurch_all["last_review"] = pd.to_datetime(
    christchurch_all["last_review"],
    errors="coerce"
)

Check duplicates

In [8]:
print("Invalid prices:",
      (christchurch_all["price"] <= 0).sum())

print("Invalid minimum nights:",
      (christchurch_all["minimum_nights"] <= 0).sum())

print("Invalid availability:",
      ((christchurch_all["availability_365"] < 0) |
       (christchurch_all["availability_365"] > 365)).sum())

Invalid prices: 0
Invalid minimum nights: 0
Invalid availability: 0


The final cleaned listing dataset

In [9]:
christchurch_clean = christchurch_all.drop(
    columns=["name", "host_name", "license"]
).copy()

In [10]:
christchurch_clean.shape

(18128, 16)

In [11]:
christchurch_clean.head()

,id,host_id,neighbourhood_group,neighbourhood,latitude,longitude,room_type,price,minimum_nights,number_of_reviews,last_review,reviews_per_month,calculated_host_listings_count,availability_365,number_of_reviews_ltm,month_year
2,1097487,4993285.0,Christchurch City,Fendalton Ward,-43.511490,172.625400,Private room,166.0,1.0,230,2025-03-03,2.17,1.0,221,14,October 2025
4,1211070132323635531,179297759.0,Christchurch City,Riccarton Ward,-43.538437,172.578953,Private room,126.0,1.0,8,2025-03-01,0.58,20.0,359,5,October 2025
5,1454660578606474393,153948924.0,Christchurch City,Fendalton Ward,-43.511080,172.623880,Entire home/apt,585.0,2.0,0,NaT,NaN,3.0,78,0,October 2025
6,52660229,425993439.0,Christchurch City,Riccarton Ward,-43.529020,172.610730,Private room,145.0,1.0,6,2025-01-11,0.26,4.0,348,2,October 2025
7,809778469981150692,497329414.0,Christchurch City,Heathcote Ward,-43.562280,172.712870,Entire home/apt,450.0,2.0,0,NaT,NaN,1.0,363,0,October 2025


In [12]:
christchurch_clean.isnull().sum()

id                                   0
host_id                              0
neighbourhood_group                  0
neighbourhood                        0
latitude                             0
longitude                            0
room_type                            0
price                                0
minimum_nights                       9
number_of_reviews                    0
last_review                       1541
reviews_per_month                 1541
calculated_host_listings_count       0
availability_365                     0
number_of_reviews_ltm                0
month_year                           0
dtype: int64

Save the cleaned Christchurch dataset

In [13]:
christchurch_clean.to_csv(
    "cleaned_christchurch_listings.csv",
    index=False
)

Document our cleaning decisions:

The Christchurch listing dataset was cleaned --> prepare to rental bond data.

The name column was removed because it identifies individual listings and is not required for comparing rental prices or property availability.

The host_name column was removed --> it is not relevant to main question

The license column was removed --> no useful--> non missing information

Rows with missing price values were removed because rental price is a key variable for the planned analysis. A missing price cannot be reliably used in price comparisons.

Rows with missing last_review and reviews_per_month were retained because these variables are not central to the main rental price and availability analysis. Missing review information can occur naturally for listings with little or no review history.

Duplicate rows were checked and removed if present.

Latitude and longitude were retained because the assignment requires them and they may be useful for future geographic analysis.

The month_year variable was retained because the dataset needs to be compared with the rental bond data over the same time period.

Document the consequences

In [14]:
initial_rows = len(christchurch_all)



final_rows = len(christchurch_clean)

print("Initial rows:", initial_rows)
print("Final rows:", final_rows)
print("Total rows removed:", initial_rows - final_rows)

Initial rows: 18128
Final rows: 18128
Total rows removed: 0


Consequences of Cleaning

Original dataset --> 28,795 rows

Rows with missing rental prices were removed because price is required for the planned rental-price comparison.

The name, host_name, and license columns were removed because they were not required for the planned analysis.

The bond dataset

In [18]:
import pandas as pd

bond = pd.read_csv(
    "Detailed quarterly report 2020-2026.xlsx.csv"
)

bond.head()

,TimeFrame,Location Id,Dwelling Type,Number Of Beds,Total Bonds,Active Bonds,Closed Bonds,Median Rent,Geometric Mean Rent,Upper Quartile Rent,Lower Quartile Rent,Log Std Dev Weekly Rent
0,2026-04-01,NaN,ALL,1,2064,19563,1176,NaN,NaN,NaN,NaN,NaN
1,2026-04-01,-99.0,ALL,1,22014,113820,6969,550.0,522.0,660.0,450.0,0.4144
2,2026-04-01,100100.0,ALL,1,6,9,0,425.0,408.0,475.0,370.0,0.2834
3,2026-04-01,100600.0,ALL,1,6,12,3,588.0,585.0,720.0,545.0,0.2550
4,2026-04-01,100700.0,ALL,1,6,30,3,550.0,530.0,579.0,473.0,0.1588


In [19]:
bond.shape

(226080, 12)

In [20]:
bond.columns

Index(['TimeFrame', 'Location Id', 'Dwelling Type', 'Number Of Beds',
       'Total Bonds', 'Active Bonds', 'Closed Bonds', 'Median Rent',
       'Geometric Mean Rent', 'Upper Quartile Rent', 'Lower Quartile Rent',
       'Log Std Dev Weekly Rent'],
      dtype='str')

In [21]:
bond.info()

<class 'pandas.DataFrame'>
RangeIndex: 226080 entries, 0 to 226079
Data columns (total 12 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   TimeFrame                226080 non-null  str    
 1   Location Id              225286 non-null  float64
 2   Dwelling Type            226080 non-null  str    
 3   Number Of Beds           214071 non-null  str    
 4   Total Bonds              226080 non-null  int64  
 5   Active Bonds             226080 non-null  int64  
 6   Closed Bonds             226080 non-null  int64  
 7   Median Rent              225277 non-null  float64
 8   Geometric Mean Rent      225277 non-null  float64
 9   Upper Quartile Rent      225277 non-null  float64
 10  Lower Quartile Rent      225277 non-null  float64
 11  Log Std Dev Weekly Rent  225221 non-null  float64
dtypes: float64(6), int64(3), str(3)
memory usage: 20.7 MB


In [22]:
bond.isnull().sum()

TimeFrame                      0
Location Id                  794
Dwelling Type                  0
Number Of Beds             12009
Total Bonds                    0
Active Bonds                   0
Closed Bonds                   0
Median Rent                  803
Geometric Mean Rent          803
Upper Quartile Rent          803
Lower Quartile Rent          803
Log Std Dev Weekly Rent      859
dtype: int64

Clean the bond dataset

In [23]:
bond.isnull().sum()

TimeFrame                      0
Location Id                  794
Dwelling Type                  0
Number Of Beds             12009
Total Bonds                    0
Active Bonds                   0
Closed Bonds                   0
Median Rent                  803
Geometric Mean Rent          803
Upper Quartile Rent          803
Lower Quartile Rent          803
Log Std Dev Weekly Rent      859
dtype: int64

In [24]:
bond.duplicated().sum()

np.int64(0)

In [25]:
bond.dtypes

TimeFrame                      str
Location Id                float64
Dwelling Type                  str
Number Of Beds                 str
Total Bonds                  int64
Active Bonds                 int64
Closed Bonds                 int64
Median Rent                float64
Geometric Mean Rent        float64
Upper Quartile Rent        float64
Lower Quartile Rent        float64
Log Std Dev Weekly Rent    float64
dtype: object

Save the cleaned bond dataset

In [30]:
bond.to_csv("cleaned_bond_data.csv",index=False)